player/lru_cache.go

```diff
@@ -0,0 +1,149 @@
+	SweepInterval time.Duration
+}
+
+type lruCache struct {
+	storage *fsStorage
+	lru     *lru.Cache
+}
+
+// InitLRUCache initializes a LRU cache for chunks.
+func InitLRUCache(opts *LRUCacheOpts) (ChunkCache, error) {
+	storage, err := initFSStorage(opts.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	if opts.Size == 0 {
+		opts.Size = defaultMaxCacheSize
+	}
+
+	onEvicted := func(key interface{}, value interface{}) {
+		storage.remove(value)
+	}
+
+	lru, err := lru.NewWithEvict(int(opts.Size/ChunkSize), onEvicted)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &lruCache{storage, lru}
+
+	Logger.Infof("LRU cache of %vGB initialized at %v", opts.Size/1024/1024/1024, opts.Path)
+
+	go func() {
+		Logger.Infoln("restoring cache in memory...")
+		err := c.reloadCache()
+		if err != nil {
+			Logger.Errorf("failed to restore cache in memory: %s", err.Error())
+		}
+	}()
+
+	return c, nil
+}
+
+func (c *lruCache) reloadCache() error {
+	err := filepath.Walk(c.storage.path, func(path string, info os.FileInfo, err error) error {
+		if c.storage.path == path {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return fmt.Errorf("subfolder %v found inside cache folder", path)
+		}
+		if len(info.Name()) != stream.BlobHashHexLength {
+			return fmt.Errorf("non-cache file found at path %v", path)
+		}
+		c.lru.Add(info.Name(), info.Name())
+		return nil
+	})
+	return err
+}
+
+func (c *lruCache) Set(hash string, body []byte) (ReadableChunk, error) {
+	var numWritten int
+	Logger.Debugf("attempting to cache chunk %v", hash)
+	chunkPath := c.storage.getPath(hash)
+
+	f, err := os.OpenFile(chunkPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0600)
+	if os.IsExist(err) {
+		MetCacheErrorCount.Inc()
+		Logger.Debugf("chunk %v already exists on the local filesystem, not overwriting", hash)
+	} else {
+		numWritten, err = f.Write(body)
+		defer f.Close()
+		if err != nil {
+			MetCacheErrorCount.Inc()
+			Logger.Errorf("error saving cache file %v: %v", chunkPath, err)
+			return nil, err
+		}
+
+		err = f.Close()
+		if err != nil {
+			MetCacheErrorCount.Inc()
+			Logger.Errorf("error closing cache file %v: %v", chunkPath, err)
+			return nil, err
+		}
+	}
+
+	evicted := c.lru.Add(hash, hash)
+	Logger.Debugf("cached chunk %v, retrieved: %vB, evicted: %v", hash, numWritten, evicted)
+
+	return &cachedChunk{reflectedChunk{body}}, nil
+}
+
+func (c *lruCache) Has(hash string) bool {
+	return c.lru.Contains(hash)
+}
+
+func (c *lruCache) Get(hash string) (ReadableChunk, bool) {
+	if value, ok := c.lru.Get(hash); ok {
+		f, err := c.storage.open(value)
+		if err != nil {
+			MetCacheErrorCount.Inc()
+			Logger.Errorf("chunk %v found in cache but couldn't open the file: %v", hash, err)
+			c.lru.Remove(value)
+			return nil, false
+		}
+		cb, err := initCachedChunk(f)
+		if err != nil {
+			Logger.Errorf("chunk %v found in cache but couldn't read the file: %v", hash, err)
+			return nil, false
+		}
+		defer f.Close()
+		return cb, true
```

Comment:
we might want to consider adding a score or something to those blobs that are most recently requested so that we don't evict them. Not sure if this is already being done somewhere else in the code, but it might be important to avoid bandwidth waste

Reply:
That's the definition of LRU, unless I'm missing something:

> Discards the least recently used items first.